### DISCUSSION TITLE
Bare minimum C++ example

### DISCUSSION THREAD

**Discussion by j-friedl** [ifm/ifm3d#358] [Q&A] [answered]
Hi,
I'm trying to build [this example](https://ifm3d.com/sphinx-doc/build/html/ifm3d/doc/sphinx/content/examples/o3r/getting_data/getting_data.html#the-full-example).

The python version works, however, I can't get the C++ example to link correctly. Is there an example Makefile or something that I could peek into? :)

edit:

I missed to provide a bit of background:
I'm using the O3R vision system and ifm3d version 0.93 on Ubuntu 20.04 (also got some docker containers running with the ifm3d library on it)

**Comment by lola-masson** [comment]
Hi @j-friedl,

You can build the examples by adding -DBUILD_EXAMPLES=ON when you build ifm3d. So:
- Go to where you cloned ifm3d (or clone it if you haven't already). 
- Make sure you are on the with branch (v0.93.0)
- Create a build folder if you haven't already
- Build with (from the build folder):
```
cmake -DBUILD_EXAMPLES=ON ..
cmake --build .
```
- The examples executables will be in build/examples/o3r

**Reply by j-friedl** [reply]
Yeah these work without a problem. However, I really don't want to compile the whole thing. Let's say I'm in my home folder and have the ifm3d library decompressed in /home/username/ifm3d/ and installed. Now I have the said example code in /home/username/example/. What would the CMakeList.txt have to look like? I'm sorry if this is obvious, I'm not really experienced with CMake.

Right now, my CMakeList.txt looks like that:

```
cmake_minimum_required(VERSION 3.1)
project(ifm3d-example)

add_executable(main src/main.cpp)
target_link_libraries(main
                      ifm3d_camera
                      ifm3d_framegrabber
                      ifm3d_stlimage
                      )
```

Where the "getting_data" example code is in src/main.cpp. When running "make", I get the following error:

```
/usr/bin/ld: CMakeFiles/main.dir/src/main.cpp.o: undefined reference to symbol '_ZN3fLI7FLAGS_vE'
/usr/bin/ld: /lib/libglog.so.0.3.5: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/main.dir/build.make:84: main] Error 1
make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/main.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

edit:
Something like [this PCL tutorial](https://pointclouds.org/documentation/tutorials/using_pcl_pcl_config.html#)  would be great in the docs.

**Comment by lola-masson** [comment]
I think you are missing the `find_package` instruction: `find_package(ifm3d REQUIRED)`.

Edit:
Good idea for the tutorial. I am adding it to the list.

**Reply by j-friedl** [reply]
Thanks, but so far, I wasn't able to make it work. Could anyone try to replicate the error/suggest a working solution?

**Reply by inbangsa** [reply]
Hello j-friedl, 

Can you please provide the CMakeLists.txt and command line cmake command that you are using to build the example along with error log that you are getting now.

**Reply by j-friedl** [reply]
`main.cpp`:

```
/*
 * Copyright 2021-present ifm electronic, gmbh
 * SPDX-License-Identifier: Apache-2.0
 */
#include <iostream>
#include <ifm3d/camera/camera_o3r.h>
#include <ifm3d/fg.h>
#include <ifm3d/stlimage.h>
#include <ifm3d/fg/distance_image_info.h>


int main(){

    //////////////////////////
    // Declare the objects:
    //////////////////////////
    // Declare the device object (one object only, corresponding to the VPU)
    auto cam = std::make_shared<ifm3d::O3RCamera>();
    // Declare the FrameGrabber and ImageBuffer objects. 
    // One FrameGrabber per camera head (define the port number).
    const auto FG_PCIC_PORT = cam->Get()["/ports/port2/data/pcicTCPPort"_json_pointer];
    auto fg = std::make_shared<ifm3d::FrameGrabber>(cam, ifm3d::DEFAULT_SCHEMA_MASK, FG_PCIC_PORT);
    auto im =  std::make_shared<ifm3d::StlImageBuffer>();

    //////////////////////////
    // Get a frame:
    //////////////////////////
    if (! fg->WaitForFrame(im.get(), 3000))
    {
      std::cerr << "Timeout waiting for camera!" << std::endl;
      return -1;
    }
    
    //////////////////////////
    // Example for 3D data:
    //////////////////////////
    auto dist = im->DistanceImage();

    std::cout << dist.height() << " " << dist.width() << std::endl;

    return 0;
}
```

`CMakeLists.txt`:

```
# Specify the minimum version
cmake_minimum_required(VERSION 3.9)
# Specify the project name
project(ifm3d-example)
# Declare the project target
add_executable(${PROJECT_NAME}
main.cpp
)

find_package(ifm3d REQUIRED)
# Libraries to link
target_link_libraries(${PROJECT_NAME}
                     ifm3d_camera
                     ifm3d_framegrabber
                     ifm3d_stlimage
                     )

```

These are in the same folder. Afterwards I do the standard cmake procedure:

> mkdir build
> cd build
> cmake ..
> make

I use VSCode on Ubuntu 20.04. The Intellisense for the ifm3d library works. Btw my installation process of the library:

1. Download .tar.gz of the tagged 0.93 version in GitHub
2. extract
3. Add VERSION file in the extracted folder with following content:

> v0.93.0+2.e724b30*v0.93.0*0*93*0**2*e724b30

4. Comment the lines 38 to 96 (included) and line 111 in the ifm3d/cmake/modules/GetVersionFromGitTag.cmake
5. mkdir build && cd build
6. cmake -DCMAKE_INSTALL_PREFIX=/usr ..
7. make
8. make install

This is the error message I get:

> [ 50%] Linking CXX executable ifm3d-example
> /usr/bin/ld: CMakeFiles/ifm3d-example.dir/main.cpp.o: undefined reference to symbol '_ZN3fLI7FLAGS_vE'
> /usr/bin/ld: /lib/libglog.so.0.3.5: error adding symbols: DSO missing from command line
> collect2: error: ld returned 1 exit status
> make[2]: *** [CMakeFiles/ifm3d-example.dir/build.make:84: ifm3d-example] Error 1
> make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/ifm3d-example.dir/all] Error 2
> make: *** [Makefile:84: all] Error 2

edit: typo, added error msg

**Comment by j-friedl** [comment, marked as answer]
I added `glog` to my library list and it compiled.

So my `CMakeLists.txt` looks like that:

```
# Specify the minimum version
cmake_minimum_required(VERSION 3.0)
# Specify the project name
project(ifm3d-example)
# Declare the project target
add_executable(${PROJECT_NAME}
main.cpp
)

find_package(ifm3d REQUIRED)
# Libraries to link
target_link_libraries(${PROJECT_NAME} PRIVATE
                     ifm3d_framegrabber
                     ifm3d_camera
                     ifm3d_stlimage
                     glog
                     )
```

After 

> mkdir build && cd build
> cmake ..
> make

I get:

> Scanning dependencies of target ifm3d-example
> [ 50%] Building CXX object CMakeFiles/ifm3d-example.dir/main.cpp.o
> [100%] Linking CXX executable ifm3d-example
> /usr/bin/ld: warning: libglog.so.0.3.5, needed by /usr/lib/gcc/x86_64-linux-gnu/9/../../../../lib/libifm3d_framegrabber.so, may conflict with libglog.so.0
> [100%] Built target ifm3d-example

When running `./buid/ifm3d-example` I get

> ERROR: flag 'logtostderr' was defined more than once (in files '/home/friedlj/ext/ifm3d-0.93.0_working/ifm3d-0.93.0/build/_deps/glog-src/src/logging.cc' and 'src/logging.cc').

So a little something is still missing. (See comment for solution)

**Reply by j-friedl** [reply]
I needed to specify the glog version as follows:

```
# Specify the minimum version
cmake_minimum_required(VERSION 3.0)
# Specify the project name
project(ifm3d-example)
# Declare the project target
add_executable(${PROJECT_NAME}
main.cpp
)

find_package(ifm3d REQUIRED)
# Libraries to link
target_link_libraries(${PROJECT_NAME} PRIVATE
                     ifm3d_framegrabber
                     ifm3d_camera
                     ifm3d_stlimage
                     /usr/lib/libglog.so.0.3.5
                     )
```

FINALLY IT WORKS xD

**Comment by inbangsa** [comment]
@j-friedl Thanks yes, glog linking was the issue,  I will check why glog lib is needed outside ifm3d as glog is internal dependency of ifm3d.
